### ISSUE TITLE
django-mptt

### ISSUE THREAD

**Issue by hardfalcon** [rdmorganiser/rdmo#368]
The maintainer of django-mptt has marked the project as unmaintained, although it looks like he is still actively developing it:
* https://github.com/django-mptt/django-mptt/blob/79301e1cb8902535c4954b32196963a6794bd74b/CHANGELOG.rst#013
* https://twitter.com/matthispam/status/1367205308439949313

The requirements.txt in rdmo 1.6-rc.1 [lists](https://github.com/rdmorganiser/rdmo/blob/1.6-rc.1/requirements.txt#L12) django-mptt 0.12 as a dependency - can this be safely updated to 0.13.2 if support for Python 3.6 is not needed, or are there any additional compatibility issues known/to be expected?

**Comment by afuetterer** [comment]
The required version is set to `django-mptt~=0.13.4` in https://github.com/rdmorganiser/rdmo/blob/c9b95607ba8880e20f573fe2cd86b46b95ccbd96/pyproject.toml#L49

The tests are passing.

The question on how to deal with unmaintained dependencies still remains relevant.

**Comment by afuetterer** [comment]
There is also some recent activity in this alternative library:
https://github.com/django-treebeard/django-treebeard/

**Comment by jochenklar** [comment]
Yes, I know of tree beard, and I would use it it I would need a tree in a new project. Replacing mptt was is not the highest priority right now. For `Attribute` we actually don't really need a tree anymore. The projects should be migrated to treebeard eventually.

**Comment by afuetterer** [comment]
Okay, so this issue should be left as is for the time being?

Long term goal: replace django-mptt -> django-treebeard

**Comment by afuetterer** [comment]
Version in https://github.com/rdmorganiser/rdmo/pull/693: `django-mptt~=0.14`
Tests are passing.

**Comment by afuetterer** [comment]
This might also be worth a look, if trees are needed:
https://github.com/fabiocaccamo/django-treenode

**Comment by MyPyDavid** [comment]
Another interesting discussion about this https://github.com/vittoriozamboni/django-groups-manager/issues/71

**Comment by jochenklar** [comment]
Besides https://github.com/django-treebeard/django-treebeard we schould look if https://github.com/feincms/django-tree-queries is a better fit.

For the attributes, we could remove the tree altogether and remove `parent` and refactor `key` to `uri_path` and query with https://docs.djangoproject.com/en/6.0/ref/models/querysets/#startswith.

**Comment by maklem** [comment]
**tl;dr**: I have looked into mptt, treebeard (Adjectancy Lists, Nested Set, Materialized Path), and tree-queries. They all do trees with different balancing between read-performance and stability during changes.

To me there is no clear best fit.
* `treebeard.ns_tree` is most similar to MPTT, but too different for a drop in replacement.
* `treebeard.mp_tree` is efficient, but needs reimplementation of some of the tree logic for migration
* `tree-queries` becomes slow for large tables.

I currently prefer `django-tree-queries` but will look into `mp_tree` once more.

---
> For the attributes, we could remove the tree altogether and remove parent and refactor key to uri_path and query with https://docs.djangoproject.com/en/6.0/ref/models/querysets/#startswith.

That is a manual version of a materialized path tree 😅 

---
## Context
I gathered some statistical data from our setup. Out of ~300 projects 4 have the parent field set.
**If you have data on this, please share**

## Theory
* Adjacency List
  * store parent node as foreign key; simple, reliable, slow
* [Nested Set Model](https://en.wikipedia.org/wiki/Nested_set_model)
  * faster reads, updating tree takes a lot of resources and care
* Materialized Path
  * no good explaining source, maybe see: https://django-treebeard.readthedocs.io/en/stable/mp_tree.html
  * path to root encoded as string, really fast reads

## Options Evaluated
### MPTT (Nested Sets + Adjectancy List as API)
[Django-MPTT](https://github.com/django-mptt/django-mptt) is most used[^use], but in "keep alive" mode (updated to new python and django versions, some bugfixes, but no features, no commitment on future updates).
 * Uses `parent` attribute as source of truth
 * builds nested set tree (+`tree_id`) as part of a nodes `.save()`, which allows fast reads of related tree nodes

### Tree Queries (Adjacency List with recursive [Common Table Expressions](https://www.postgresql.org/docs/current/queries-with.html))
* Good for frequently changing trees, (or non-tree nodes: see benchmark)
* slow with trees in large tables

### Treebeard
Implementation of all mentioned strategies.

Materialized Paths are really efficient for reads, but migrating to them/from them is far from trivial. Treebeard does not provide migration helpers.

## Benchmark (get ancestors)

For testing I create a table with 500 Non-Tree-Nodes, and 20 trees with 4 one-node-layers. I then pick a node (tree or non-tree) and benchmark the query for it and its ancestors.

code temporarily at https://github.com/maklem/temp-tree-implementations-benchmark

```
------------------------------------------------------------------------------------------------ benchmark: 10 tests -------------------------------------------------------------------------------------------------
Name (time in us)                        Min                   Max                  Mean                StdDev                Median                 IQR            Outliers         OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_beard_mp[nontree-123]          258.2120 (1.0)      1,505.3010 (1.41)       768.5905 (1.38)       344.6449 (1.42)       663.5775 (1.50)     527.4140 (1.77)         14;0  1,301.0829 (0.72)         50           1
test_queries[nontree-123]           287.0450 (1.11)     1,064.8040 (1.0)        556.6241 (1.0)        243.2174 (1.0)        443.4675 (1.0)      423.8010 (1.43)         15;0  1,796.5444 (1.0)          50           1
test_beard_mp[root-4-child-2]       300.0700 (1.16)     1,510.1360 (1.42)       854.8793 (1.54)       312.9177 (1.29)       904.1705 (2.04)     428.7990 (1.44)         15;0  1,169.7557 (0.65)         50           1
test_mptt[nontree-123]              306.8220 (1.19)     1,914.7890 (1.80)     1,023.2782 (1.84)       419.7254 (1.73)       986.1355 (2.22)     742.9550 (2.50)         20;0    977.2514 (0.54)         50           1
test_beard_ns[nontree-123]          522.8950 (2.03)     2,264.8080 (2.13)       977.1823 (1.76)       381.7010 (1.57)       882.7155 (1.99)     406.5590 (1.37)         13;4  1,023.3505 (0.57)         50           1
test_beard_al[nontree-123]          532.0320 (2.06)     1,773.6510 (1.67)     1,011.4887 (1.82)       299.7534 (1.23)       987.1305 (2.23)     297.2040 (1.0)          18;3    988.6418 (0.55)         50           1
test_beard_ns[root-4-child-2]       780.5250 (3.02)     9,778.0370 (9.18)     1,685.5223 (3.03)     1,399.8889 (5.76)     1,325.0550 (2.99)     549.2640 (1.85)          3;5    593.2879 (0.33)         50           1
test_mptt[root-4-child-2]           990.9680 (3.84)     4,858.0470 (4.56)     2,075.2007 (3.73)       702.9756 (2.89)     1,967.0145 (4.44)     852.2700 (2.87)         12;1    481.8811 (0.27)         50           1
test_queries[root-4-child-2]      1,516.8280 (5.87)     5,724.3920 (5.38)     3,909.7890 (7.02)       769.3714 (3.16)     3,957.8340 (8.92)     982.6220 (3.31)         10;2    255.7683 (0.14)         50           1
test_beard_al[root-4-child-2]     2,121.6450 (8.22)     4,320.3350 (4.06)     2,992.9727 (5.38)       531.7987 (2.19)     2,929.6975 (6.61)     831.3510 (2.80)         19;0    334.1160 (0.19)         50           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

My summary: Tree-Queries are really fast, if there is no tree, but painfully slow with a tree. Not visible in this snapshot: tree-queries becomes slower for more elements in the table, including ones not part of the current tree.


[^use]: https://djangopackages.org/grids/g/trees-and-graphs/

**Comment by maklem** [comment]
Currently the most advanced query I know of is
* For every unique project (viewsets.py:161ff, managers.py:15ff)
  * in a user's projects
    * and any descendant
* get the user's highest membership rank (subquery in viewsets.py:181ff)
  * within the project
    *  and its ancestors  

I do not see that work out as single query with MP or AL trees. As far as I can tell, only treebeard.ns_tree is left. (there is django-polymorphic-tree, but it's less maintained than django-mptt)

<details>
<summary>mentioned code</summary>

https://github.com/rdmorganiser/rdmo/blob/2dea2d979d9d00a9de9b44961420963bc0fbcab4/rdmo/projects/viewsets.py#L148-L217

https://github.com/rdmorganiser/rdmo/blob/2dea2d979d9d00a9de9b44961420963bc0fbcab4/rdmo/projects/managers.py#L15-L44
</details>
<details>
<summary>An example of this query extracted from tests</summary>

```sql
SELECT DISTINCT "projects_project"."id", "projects_project"."created", "projects_project"."updated", "projects_project"."parent_id", "projects_project"."site_id", "projects_project"."title", "projects_project"."description", "projects_project"."catalog_id", "projects_project"."progress_total", "projects_project"."progress_count", "projects_project"."lft", "projects_project"."rght", "projects_project"."tree_id", "projects_project"."level", 
(
    SELECT U0."role" AS "role" FROM "projects_membership" U0 
    INNER JOIN "projects_project" U1 ON (U0."project_id" = U1."id") 
    WHERE (U0."project_id" = ("projects_project"."id") AND U0."user_id" = 4) ORDER BY U1."title" ASC
) AS "current_role",
(
    SELECT U0."role" AS "role" FROM "projects_membership" U0 
    INNER JOIN "projects_project" U1 ON (U0."project_id" = U1."id") 
    WHERE (U1."lft" <= ("projects_project"."lft") AND U1."rght" >= ("projects_project"."rght") AND U1."tree_id" = ("projects_project"."tree_id") AND U0."user_id" = 4) 
    ORDER BY CASE 
        WHEN (U0."role" = \'owner\') THEN 3 
        WHEN (U0."role" = \'manager\') THEN 2 
        WHEN (U0."role" = \'author\') THEN 1 
        WHEN (U0."role" = \'guest\') THEN 0 
        ELSE NULL END DESC LIMIT 1
) AS "highest_role",
(
    SELECT U0."project_id" AS "project_id" FROM "projects_membership" U0 
    INNER JOIN "projects_project" U1 ON (U0."project_id" = U1."id") 
    WHERE (U1."lft" <= ("projects_project"."lft") AND U1."rght" >= ("projects_project"."rght") AND U1."tree_id" = ("projects_project"."tree_id") AND U0."user_id" = 4) 
    ORDER BY CASE WHEN (U0."role" = \'owner\') THEN 3 WHEN (U0."role" = \'manager\') THEN 2 WHEN (U0."role" = \'author\') THEN 1 WHEN (U0."role" = \'guest\') THEN 0 ELSE NULL END DESC LIMIT 1
) AS "highest_role_project_id",
(
    SELECT U1."title" AS "project__title" FROM "projects_membership" U0 
    INNER JOIN "projects_project" U1 ON (U0."project_id" = U1."id") 
    WHERE (U1."lft" <= ("projects_project"."lft") AND U1."rght" >= ("projects_project"."rght") AND U1."tree_id" = ("projects_project"."tree_id") AND U0."user_id" = 4) 
    ORDER BY CASE WHEN (U0."role" = \'owner\') THEN 3 WHEN (U0."role" = \'manager\') THEN 2 WHEN (U0."role" = \'author\') THEN 1 WHEN (U0."role" = \'guest\') THEN 0 ELSE NULL END DESC LIMIT 1
) AS "highest_role_project_title",
(
    SELECT U0."id" AS "id" FROM "projects_membership" U0 
    INNER JOIN "projects_project" U1 ON (U0."project_id" = U1."id") 
    WHERE (U1."lft" <= ("projects_project"."lft") AND U1."rght" >= ("projects_project"."rght") AND U1."tree_id" = ("projects_project"."tree_id") AND U0."user_id" = 4) 
    ORDER BY CASE WHEN (U0."role" = \'owner\') THEN 3 WHEN (U0."role" = \'manager\') THEN 2 WHEN (U0."role" = \'author\') THEN 1 WHEN (U0."role" = \'guest\') THEN 0 ELSE NULL END DESC LIMIT 1
) AS "highest_role_membership_id", 
COALESCE(
    MAX(
        (
            SELECT U0."updated" AS "updated" FROM "projects_value" U0 
            WHERE U0."project_id" = ("projects_project"."id") 
            ORDER BY 1 DESC LIMIT 1
        ), "projects_project"."updated"
    ), "projects_project"."updated"
) AS "last_changed",
"questions_catalog"."id", "questions_catalog"."created", "questions_catalog"."updated", "questions_catalog"."uri", "questions_catalog"."uri_prefix", "questions_catalog"."uri_path", "questions_catalog"."comment", "questions_catalog"."locked", "questions_catalog"."order", "questions_catalog"."title_lang1", "questions_catalog"."title_lang2", "questions_catalog"."title_lang3", "questions_catalog"."title_lang4", "questions_catalog"."title_lang5", "questions_catalog"."help_lang1", "questions_catalog"."help_lang2", "questions_catalog"."help_lang3", "questions_catalog"."help_lang4", "questions_catalog"."help_lang5", "questions_catalog"."available", "projects_visibility"."id", "projects_visibility"."created", "projects_visibility"."updated", "projects_visibility"."project_id" FROM "projects_project" 
LEFT OUTER JOIN "projects_membership" ON ("projects_project"."id" = "projects_membership"."project_id") 
LEFT OUTER JOIN "projects_visibility" ON ("projects_project"."id" = "projects_visibility"."project_id") 
LEFT OUTER JOIN "projects_visibility_sites" ON ("projects_visibility"."id" = "projects_visibility_sites"."visibility_id") 
LEFT OUTER JOIN "projects_visibility_groups" ON ("projects_visibility"."id" = "projects_visibility_groups"."visibility_id") 
INNER JOIN "projects_membership" T9 ON ("projects_project"."id" = T9."project_id") 
LEFT OUTER JOIN "questions_catalog" ON ("projects_project"."catalog_id" = "questions_catalog"."id") 
WHERE (
    (
        "projects_membership"."user_id" = 4 OR (
            "projects_visibility"."id" IS NOT NULL AND (
                "projects_visibility_sites"."site_id" IS NULL OR "projects_visibility_sites"."site_id" = 1
            ) AND (
                "projects_visibility_groups"."group_id" IS NULL OR "projects_visibility_groups"."group_id" IN (
                    SELECT U0."id" FROM "auth_group" U0 INNER JOIN "auth_user_groups" U1 ON (U0."id" = U1."group_id") WHERE U1."user_id" = 4
                )
            )
        )
    ) AND T9."user_id" = 5
) 
ORDER BY 20 DESC 
LIMIT 1
```
</details>
